pageserver/src/tenant/vectored_blob_io.rs

```diff
@@ -144,6 +247,107 @@ impl VectoredReadBuilder {
+    start_blk_no: usize,
+    /// End block number (exclusive).
+    end_blk_no: usize,
+    /// Blobs (metadata, end offset) ordered by blobs start offset.
+    blobs_at: VecMap<u64, (u64, BlobMeta)>,
+    max_read_size: Option<usize>,
+    /// Chunk size reads are coalesced into.
+    chunk_size: usize,
+}
+
+/// Computes x / d rounded up.
+fn div_round_up(x: usize, d: usize) -> usize {
+    (x + (d - 1)) / d
+}
+
+impl ChunkedVectoredReadBuilderInner {
+    /// Start building a new vectored read.
+    ///
+    /// Note that by design, this does not check against reading more than `max_read_size` to
+    /// support reading larger blobs than the configuration value. The builder will be single use
+    /// however after that.
+    pub(crate) fn new(
+        start_offset: u64,
+        end_offset: u64,
+        meta: BlobMeta,
+        max_read_size: Option<usize>,
+        chunk_size: usize,
+    ) -> Self {
+        let mut blobs_at = VecMap::default();
+        blobs_at
+            .append(start_offset, (end_offset, meta))
+            .expect("First insertion always succeeds");
+
+        let start_blk_no = start_offset as usize / chunk_size;
+        let end_blk_no = div_round_up(end_offset as usize, chunk_size);
+        Self {
+            start_blk_no,
+            end_blk_no,
+            blobs_at,
+            max_read_size,
+            chunk_size,
+        }
+    }
+
+    /// Attempts to extend the current read with a new blob if the new blob resides in the same or the immediate next chunk.
+    ///
+    /// The resulting size also must be below the max read size.
+    pub(crate) fn extend(&mut self, start: u64, end: u64, meta: BlobMeta) -> VectoredReadExtended {
+        tracing::trace!(start, end, "trying to extend");
+        let start_blk_no = start as usize / self.chunk_size;
+        let end_blk_no = div_round_up(end as usize, self.chunk_size);
+
+        let not_limited_by_max_read_size = || {
+            if let Some(max_read_size) = self.max_read_size {
+                let coalesced_size = (end_blk_no - self.start_blk_no) * self.chunk_size;
+                coalesced_size <= max_read_size
+            } else {
+                true
+            }
+        };
+
+        /// Returns true if the second block starts in the same block or the immediate next block where the first block ended.
+        ///
+        /// Note: This automatically handles the case where two blocks are adjacent to each other,
+        /// whether they starts on chunk size boundary or not.
+        fn is_adjacent_chunk_read(first_end_blk_no: usize, second_start_blk_no: usize) -> bool {
+            // 1. first.end & second.start are in the same block
+            first_end_blk_no == second_start_blk_no + 1 ||
+            // 2. first.end ends one block before second.start
+            first_end_blk_no == second_start_blk_no
+        }
+
+        if is_adjacent_chunk_read(self.end_blk_no, start_blk_no) && not_limited_by_max_read_size() {
```

Comment:
Style: why is this an function? It's called exactly once.

Why not

```
let is_adjacent_chunk_read = { ...};

if is_adjacent_chunk_read && not_limited_by_max_read_size { ... }
```

Reply:
Agreed. 9213dc31e60af84da35531598b462fb9db1cc9b5